Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/survey/SurveySettings.kt

```diff
@@ -1,78 +1,88 @@
+import de.rki.coronawarnapp.util.serialization.BaseJackson
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
 import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class SurveySettings @Inject constructor(
-    @AppContext val context: Context,
-    @BaseGson val gson: Gson
+    @SurveySettingsDataStore private val dataStore: DataStore<Preferences>,
+    @BaseJackson private val objectMapper: ObjectMapper,
 ) : Resettable {
 
-    private val preferences by lazy {
-        context.getSharedPreferences("survey_localdata", Context.MODE_PRIVATE)
+    val oneTimePassword = dataStore.dataRecovering.distinctUntilChanged(
+        key = KEY_OTP
+    ).map { json ->
+        if (json != null && json != String()) {
+            val otp = objectMapper.readValue<OneTimePassword>(json)
+            requireNotNull(otp.uuid)
+            requireNotNull(otp.time)
+            otp
+        } else {
+            null
+        }
     }
 
-    var oneTimePassword: OneTimePassword?
-        get() {
-            try {
-                val json = preferences.getString(KEY_OTP, null)
-                if (json != null) {
-                    val otp = gson.fromJson(json, OneTimePassword::class.java)
-                    requireNotNull(otp.uuid)
-                    requireNotNull(otp.time)
-                    return otp
-                }
-                return null
-            } catch (t: Throwable) {
-                Timber.e(t, "failed to parse OTP from preferences")
-                return null
-            }
+    suspend fun updateOneTimePassword(otp: OneTimePassword?) = dataStore.edit {
+        it[KEY_OTP] = when (otp) {
+            null -> String()
+            else -> objectMapper.writeValueAsString(otp)
         }
-        set(value) =
-            preferences
-                .edit()
-                .putString(KEY_OTP, if (value == null) null else gson.toJson(value))
-                .apply()
+    }
 
-    var otpAuthorizationResult: OTPAuthorizationResult?
-        get() {
-            try {
-                val json = preferences.getString(KEY_OTP_RESULT, null)
-                if (json != null) {
-                    val result = gson.fromJson(json, OTPAuthorizationResult::class.java)
-                    requireNotNull(result.uuid)
-                    requireNotNull(result.authorized)
-                    requireNotNull(result.redeemedAt)
-                    requireNotNull(result.invalidated)
-                    return result
-                }
-                return null
-            } catch (t: Throwable) {
-                Timber.e(t, "failed to parse OTP from preferences")
-                return null
+    val otpAuthorizationResult: Flow<OTPAuthorizationResult?> = dataStore.dataRecovering.distinctUntilChanged(
+        key = KEY_OTP_RESULT
+    ).map { json ->
+        try {
+            if (json != null && json != String()) {
```

Comment:
`!json.isNullOrEmpty()`

Reply:
Made nullable, so empty check is removed